File: solutions/scala/prime-factors/src/main/scala/PrimeFactors.scala
import scala.annotation.tailrec

object PrimeFactors:
  def factors(n: Long): List[Long] =
    def nextPrime(current: Long): Long =
      if current == 2 then 3 else current + 2

    @tailrec
    def iter(number: Long, prime: Long, acc: List[Long]): List[Long] =
      if number == 1 then acc
      else if prime * prime > number then number :: acc
      else if number % prime == 0 then iter(number / prime, prime, prime :: acc)
      else iter(number, nextPrime(prime), acc)

    iter(n, 2L, List()).reverse

File: solutions/go/prime-factors/prime_factors.go
package prime

import "math"

func Factors(number int64) []int64 {
	var factors []int64

	for number&1 == 0 {
		factors = append(factors, 2)
		number /= 2
	}

	max := int64(math.Sqrt(float64(number)))
	for i := int64(3); i <= max; i += 2 {
		for number%i == 0 {
			factors = append(factors, i)
			number /= i
		}
	}

	if number > 2 {
		factors = append(factors, number)
	}
	return factors
}
